app/models/node.rb

```diff
@@ -756,66 +756,68 @@ def add_barnstar(tagname, giver)
-          end
-          # add sub-tags:
-          subtags = {}
-          subtags['pm'] = 'particulate-matter'
-          if subtags.include?(key)
-            add_tag(subtags[key], user)
-          end
-          # parse date tags:
-          if key == 'date'
-            begin
-              DateTime.strptime(value, '%m-%d-%Y').to_date.to_s(:long)
-            rescue StandardError
-              return [false, tag.destroy]
+        ActiveRecord::Base.transaction do
+          if tag.valid?
+            key = tag.name.split(':')[0]
+            value = tag.name.split(':')[1]
+            # add base tags:
+            if ['question', 'upgrade', 'activity'].include?(key)
+              add_tag(value, user)
             end
-          end
-          tag.save!
-          node_tag = NodeTag.new(tid: tag.id,
+            # add sub-tags:
+            subtags = {}
+            subtags['pm'] = 'particulate-matter'
+            if subtags.include?(key)
+              add_tag(subtags[key], user)
+            end
+            # parse date tags:
+            if key == 'date'
+              begin
+                DateTime.strptime(value, '%m-%d-%Y').to_date.to_s(:long)
+              rescue StandardError
+                return [false, tag.destroy]
+              end
+            end
+            tag.save!
+            node_tag = NodeTag.new(tid: tag.id,
                                  uid: user.uid,
                                  date: DateTime.now.to_i,
                                  nid: id)
 
-          # Adding lat/lon values into node table
-          if key == 'lat'
-            tagvalue = value
-            table_updated = update_attributes(latitude: tagvalue, precision: decimals(tagvalue).to_s)
-          elsif key == 'lon'
-            tagvalue = value
-            table_updated = update_attributes(longitude: tagvalue)
-          end
+            # Adding lat/lon values into node table
+            if key == 'lat'
+              tagvalue = value
+              table_updated = update_attributes(latitude: tagvalue, precision: decimals(tagvalue).to_s)
+            elsif key == 'lon'
+              tagvalue = value
+              table_updated = update_attributes(longitude: tagvalue)
+            end
 
-          if node_tag.save
-            saved = true
-            tag.run_count # update count of tag usage
-            # send email notification if there are subscribers, status is OK, and less than 1 month old
-            unless tag.subscriptions.empty? || status == 3 || status == 4 || created < (DateTime.now - 1.month).to_i
-              SubscriptionMailer.notify_tag_added(self, tag, user).deliver_now
+            if node_tag.save
+              saved = true
+              tag.run_count # update count of tag usage
+              # send email notification if there are subscribers, status is OK, and less than 1 month old
+              unless tag.subscriptions.empty? || status == 3 || status == 4 || created < (DateTime.now - 1.month).to_i
```

Comment:
Hi @sssash18. Good job, just one more touch and it's perfect.
When I look at line 810, it is hard to evaluate what it does when you have thing1 || thing2 < thing3 && thing4...
I know it's not your code, but it would be valuable if you could refactor it.

For example:
```ruby
isStatusValid = staus == 3 || status == 4
isMonthOld = created < (DateTime.now - 1.month).to_i

unless tag.subscriptions.empty? || isStatusValid || !isMonthOld
    # ...
```

Thanks.

Reply:
@VladimirMikulic is this resolved?

Reply:
Definitely yes.